File: src/controllers/services/simple-transaction-service.ts
import SimpleTransactionRepository from '../../infra/simple-transaction-repository'
import TransferRepository from '../../infra/transfer-repository'
import RecurrenceRepository from '../../infra/recurrence-repository'

import SimpleTransaction from '../../models/simple-transaction'
import Recurrence from '../../models/recurrence'

import Transaction from '../../models/transaction'
import Transfer from '../../models/transfer'

export default class SimpleTransactionService {

    private simpleTransactionRepository: SimpleTransactionRepository = new SimpleTransactionRepository()

    async getAllActive(): Promise<SimpleTransaction[]> {
        const simpleTransactions = await this.simpleTransactionRepository.getAll()        

        const simpleTransactionsActive = simpleTransactions.filter(simpleTransaction => simpleTransaction.isActive)
        return simpleTransactionsActive
    }

    async getByIdActive(id: number): Promise<SimpleTransaction> {
        if (id === undefined || id === null) {
            throw new Error('idType is required')
        }
            const simpleTransactionEntity = await this.simpleTransactionRepository.getById(id)
            if (!simpleTransactionEntity) {
                throw new Error(`SimpleTransaction with id ${id} not found`)
            }
            if (!simpleTransactionEntity.isActive) {
                throw new Error(`SimpleTransaction with id ${id} is not active`)
            }
            return simpleTransactionEntity
    }

    async upsert(transactionRequest: SimpleTransaction) : Promise<SimpleTransaction> {

        const isCreate = transactionRequest.id === undefined || transactionRequest.id === null

        if (isCreate) {
            var simpleTransactionCreating = new SimpleTransaction(
                transactionRequest.date, 
                transactionRequest.description, 
                transactionRequest.accountId, 
                transactionRequest.amount,
                transactionRequest.isActive,
                transactionRequest.isRecurrent,
                null
            )

            const idCreate = await this.simpleTransactionRepository.create(simpleTransactionCreating)

            return simpleTransactionCreating             
        } else {
            var simpleTransactionUpdating = new SimpleTransaction(
                transactionRequest.date, 
                transactionRequest.description, 
                transactionRequest.accountId, 
                transactionRequest.amount,
                transactionRequest.isActive,
                transactionRequest.isRecurrent,
                transactionRequest.id
            )
            
            await this.simpleTransactionRepository.update(simpleTransactionUpdating)

            return simpleTransactionUpdating
        }
    }

    async delete(id: number) : Promise<void> {
        const simpleTransaction = await this.simpleTransactionRepository.getById(id)
        if (!simpleTransaction) {
            throw new Error(`SimpleTransaction with id ${id} not found`)
        }
        simpleTransaction.isActive = false
        await this.simpleTransactionRepository.update(simpleTransaction)
    }
}

File: Fin.Api/Services/SimpleTransactionService.cs
using Fin.Api.Data;
using Fin.Api.Models;
using Fin.Api.Repository;
using System.Threading.Tasks;

namespace Fin.Api.Services;

public class SimpleTransactionService(ISimpleTransactionRepository repository)
{
    private readonly ISimpleTransactionRepository _repository = repository;

    public List<SimpleTransaction> GetAllActive()
    {
        return _repository.GetAll();
    }

    public Transaction Upsert(Transaction transactionRequest)
    {
        SimpleTransaction simpleTransactionUpserting;

        var isCreate = transactionRequest.Id == 0 || transactionRequest.Id == null;

        if (isCreate)
        {
            simpleTransactionUpserting = new SimpleTransaction
            {
                Date = transactionRequest.Date,
                Description = transactionRequest.Description,
                AccountId = transactionRequest.RefAccountId,
                Amount = transactionRequest.Amount,
                IsRecurrent = transactionRequest.IsRecurrent,
                IsActive = transactionRequest.IsActive
            };

            _repository.Create(simpleTransactionUpserting);
        }
        else
        {
            simpleTransactionUpserting = new SimpleTransaction
            {
                Id = transactionRequest.Id.Value,
                Date = transactionRequest.Date,
                Description = transactionRequest.Description,
                AccountId = transactionRequest.RefAccountId,
                Amount = transactionRequest.Amount,
                IsRecurrent = transactionRequest.IsRecurrent,
                IsActive = transactionRequest.IsActive
            };
            
            _repository.Update(simpleTransactionUpserting);            
        }

        return new Transaction
        {
            Id = simpleTransactionUpserting.Id,
            Date = transactionRequest.Date,
            Description = transactionRequest.Description,
            RefAccountId = transactionRequest.RefAccountId,
            Amount = transactionRequest.Amount,
            Type = "SIMPLE",
            OtherAccountId = transactionRequest.OtherAccountId,
            IsRecurrent = transactionRequest.IsRecurrent,
            IsActive = transactionRequest.IsActive
        };
    }

    public void Delete(int id)
    {
        _repository.Delete(id);
    }
}
